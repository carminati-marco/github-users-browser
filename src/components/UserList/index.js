import React, { Component } from "react";
import { connect } from "react-redux";
import { ListGroup, InputGroup, FormControl } from "react-bootstrap";
import styled from "styled-components";

import { setUser } from "../../redux/actions";
import { getUsers } from "../../services";

const StyledFormControl = styled(FormControl)`
  border-radius: 25px;
`;

class UserList extends Component {
  state = { users: [], selectedUserId: null, search: null };

  async componentDidMount() {
    const users = await getUsers();
    this.setState({ users });
  }

  async handleChange(event) {
    const search = event.target.value;
    const users = await getUsers(search);
    this.setState({ search, users, selectedUserId: null });
    this.props.setUser(null);
  }

  handleClickListItem(event) {
    const selectedUserId = this.state.selectedUserId === event ? null : event;
    this.setState({ selectedUserId });

    const user = this.state.users.find(user => user.id === selectedUserId);
    this.props.setUser(user || null);
  }

  render() {
    const { users = [], selectedUserId } = this.state;
    return (
      <div>
        <InputGroup className="mb-3">
          <StyledFormControl
            placeholder="Search.."
            onChange={event => this.handleChange(event)}
            aria-label="Search"
            aria-describedby="basic-addon1"
          />
        </InputGroup>
        <ListGroup activeKey={selectedUserId}>
          {users.map(user => (
            <ListGroup.Item onClick={() => this.handleClickListItem(user.id)} key={user.id} eventKey={user.id}>
              {user.login}
            </ListGroup.Item>
          ))}
        </ListGroup>
      </div>
    );
  }
}

export default connect(
  null,
  { setUser }
)(UserList);
