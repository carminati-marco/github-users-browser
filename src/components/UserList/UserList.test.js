import React from "react";
import { Provider } from "react-redux";

import store from "../../redux/store";
import UserList from "./index";
import { mount } from "enzyme";

describe("UserList test", () => {
  let component;

  beforeEach(() => {
    component = mount(
      <Provider store={store}>
        <UserList />
      </Provider>
    );
  });

  it("renders correctly", () => {
    expect(component).toMatchSnapshot();
  });

  it("ListGroupItem must be 3", () => {
    const c = component.find(UserList).find("ListGroupItem").length;
    expect(c).toEqual(3);
  });

  it("ListGroupItem change using with input", () => {
    component
      .find("input")
      .first()
      .simulate("change", { target: { value: "tn" } });

    component.update();
    // checks there is only a ListGroupItem with the right id (1455367)
    expect(component.find("ListGroupItem").length).toEqual(1);
    expect(component.find("ListGroupItem div[data-rb-event-key='1455367']").length).toEqual(1);
  });

  it("ListGroupItem change using with input UPPERCASE", () => {
    component
      .find("input")
      .first()
      .simulate("change", { target: { value: "TN" } });

    component.update();
    // checks there is only a ListGroupItem with the right id (1455367)
    expect(component.find("ListGroupItem").length).toEqual(1);
    expect(component.find("ListGroupItem div[data-rb-event-key='1455367']").length).toEqual(1);
  });

  it("ListGroupItem toggle activate", () => {
    component
      .find("ListGroupItem")
      .first()
      .simulate("click");

    component.update();
    expect(component.find("ListGroupItem div[className='list-group-item active']").length).toEqual(1);

    component
      .find("ListGroupItem")
      .first()
      .simulate("click");
    expect(component.find("ListGroupItem div[className='list-group-item active']").length).toEqual(0);
  });
});
