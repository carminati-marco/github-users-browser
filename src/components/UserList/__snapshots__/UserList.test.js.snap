// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList test renders correctly 1`] = `ReactWrapper {}`;
